ring/consul_client.go

```diff
@@ -18,18 +20,20 @@ const (
 // such as CAS and Watch which take callbacks.  It also deals with serialisation
 // by having an instance factory passed in to methods and deserialising into that.
 type ConsulClient interface {
-	Get(key string, factory InstanceFactory) error
-	CAS(key string, factory InstanceFactory, f CASCallback) error
-	WatchPrefix(path string, factory InstanceFactory, done <-chan struct{}, f func(string, interface{}) bool)
-	WatchKey(key string, factory InstanceFactory, done <-chan struct{}, f func(interface{}) bool)
+	CAS(key string, f CASCallback) error
+	WatchPrefix(path string, done <-chan struct{}, f func(string, interface{}) bool)
+	WatchKey(key string, done <-chan struct{}, f func(interface{}) bool)
 	PutBytes(key string, buf []byte) error
 }
 
 // CASCallback is the type of the callback to CAS.  If err is nil, out must be non-nil.
 type CASCallback func(in interface{}) (out interface{}, retry bool, err error)
 
-// InstanceFactory type creates empty instances for use when deserialising
-type InstanceFactory func() interface{}
+// SerDes allows the consult client to serialise and deserialise values.
+type SerDes interface {
```

Comment:
`Codec` would be a more traditional name, and would be more consistent with the methods on json & proto.